proposals/shared-everything-threads/Overview.md

```diff
@@ -702,6 +702,37 @@ models. For more discussion: [How should toolchains apply `shared`?][toolchain-s
 
 [toolchain-shared-discussion]: https://github.com/abrown/thread-spawn/discussions/5
 
+### How will GCs and engines need to change?
+
+GCs that historically assume that there is only a single mutator thread will have to newly be able
+to handle heaps with multiple mutators running in parallel. Since sharedness is part of static
+types, engines can know at allocation or access time when they need to do something different for
+shared objects, so it is possible for them to add a separate concept of shared heap on top of their
+existing single-threaded heaps. This might be a reasonable intermediate architecture, even for
+engines that plan on eventually switching to having shared heaps as their only primitive.
+
+For engines that choose to have both shared and unshared heaps, references from unshared objects to
+shared objects will have to be treated as roots when the garbage collection runs on the shared heap.
+
+References from shared objects to unshared are normally disallowed, but we could choose to support
+them in limited cases to make TLS or thread-bound data more ergonomic. If we do, there are varying
+guarantees we could choose to make about how the GC treats these references.
+
+The weakest semantics would be that shared-to-unshared references never suffice to keep the unshared
+object alive. Users would have to root unshared objects for the duration of their intended
+lifetimes, so this is no more expressive than simply disallowing shared-to-unshared references.
+
+The strongest semantics would be that shared-to-unshared references behave exactly like normal
+references with respect to garbage collection. This requires the garbage collector to have
+visibility over the shared and unshared heaps from all threads so it can find unreachable cross-heap
+cycles.
+
+An intermediate semantics would be the same as the strong semantics, but without the guarantee that
```

Comment:
Being pedantic, "semantics" sounds like liveness guarantees. AFAIU there is never the inverse guarantee that something is definitely collected. It is always compliant for an implementation to leak the world -- after all, it's up to the implementation when to schedule GCs. Weak semantics means a shared-to-unshared reference does not keep the unshared object live. Strong means that a reference does keep the unshared object live. In this framing the intermediate semantics is indistinguishable from the strong semantics.

How does WasmGC formally talk about liveness? For completeness, JS has [this funny undecidable quantification](https://tc39.es/ecma262/#sec-liveness).

Reply:
WasmGC does not say anything at all about garbage collection except that the implementation may impose a limit on the number of allocated structs and arrays and terminate the execution if it is exceeded.

This is a discussion of properties that are outside the universe of properties the spec can say anything about, but I'm open to suggestions for clearer language. Perhaps "behavior" instead of "semantics"?

Reply:
"Behavior" is probably fine. The important thing is to distinguish when you're talking about prescribed (or proscribed) behavior, or practical implications for implementations.